homeassistant/components/simplisafe/__init__.py

```diff
@@ -0,0 +1,139 @@
+"""
+Support for SimpliSafe alarm systems.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/simplisafe/
+"""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.alarm_control_panel import (
+    SCAN_INTERVAL as DEFAULT_SCAN_INTERVAL)
+from homeassistant.config_entries import SOURCE_IMPORT
+from homeassistant.const import (
+    CONF_CODE, CONF_PASSWORD, CONF_SCAN_INTERVAL, CONF_TOKEN, CONF_USERNAME)
+from homeassistant.core import callback
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.event import async_track_time_interval
+
+from homeassistant.helpers import config_validation as cv
+
+from .config_flow import configured_instances
+from .const import DATA_CLIENT, DOMAIN, TOPIC_UPDATE
+
+REQUIREMENTS = ['simplisafe-python==3.1.6']
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_LISTENER = 'listener'
+
+CONF_ACCOUNTS = 'accounts'
+
+ACCOUNT_CONFIG_SCHEMA = vol.Schema({
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_CODE): cv.string,
+    vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL):
+        cv.time_period
+})
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_ACCOUNTS):
+            vol.All(cv.ensure_list, [ACCOUNT_CONFIG_SCHEMA]),
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+@callback
+def _async_save_refresh_token(hass, config_entry, token):
+    hass.config_entries.async_update_entry(
+        config_entry,
+        data={
+            CONF_USERNAME: config_entry.data[CONF_USERNAME],
+            CONF_TOKEN: token
+        })
+
+
+async def async_setup(hass, config):
+    """Set up the SimpliSafe component."""
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][DATA_CLIENT] = {}
+    hass.data[DOMAIN][DATA_LISTENER] = {}
+
+    if DOMAIN not in config:
+        return True
+
+    conf = config[DOMAIN]
+
+    for account in conf[CONF_ACCOUNTS]:
+        if account[CONF_USERNAME] in configured_instances(hass):
+            continue
+
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={'source': SOURCE_IMPORT},
+                data={
+                    CONF_USERNAME: account[CONF_USERNAME],
+                    CONF_PASSWORD: account[CONF_PASSWORD],
+                    CONF_CODE: account.get(CONF_CODE),
+                }))
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up SimpliSafe as config entry."""
+    from simplipy import API
+    from simplipy.errors import SimplipyError
+
+    websession = aiohttp_client.async_get_clientsession(hass)
+
+    try:
+        simplisafe = await API.login_via_token(
+            config_entry.data[CONF_TOKEN], websession)
+    except SimplipyError as err:
```

Comment:
Are you able to distinguish between connection not available and token invalid? On connection not available you should `raise ConfigEntryNotReady`

Reply:
I can, but check out my latest commit and see if you're okay with the method. Since I handle all token regeneration in the library, I don't raise a particular exception for it; rather, I can check the message for it.